README
ImgBrowz0r
==========

ImgBrowz0r is a class written in PHP5 (tested with 5.2.9) that enables you to set up
a simple gallery with basic configuration in 10 minutes. It's also possible to integrate
it into a CMS or your existing website. You can also configure various other things like
sorting order and thumbnail sizes (etc. etc.).

It does things like generating thumbnails, generating pagination/breadcrumbs and outputs
XHTML that has sufficient classes and ID's to make it easy to adjust and extend everything
with CSS and Javascript.

The only thing you have to do is install it and upload your images and you're done. And
maybe clean the cache once in a while.

Features
========

 - Infinite nesting of directories.
 - Automatic thumbnail creation and caching.
 - Thumbnail is refreshed if the orginal image is changed.
 - Transparency in PNG and GIF images is preserved in the thumbnail.
 - Valid XHTML 1.0 Strict output.
 - Easy to integrate into any website or CMS.
 - Easy to style and extend with CSS and Javascript.
 - Ability to configure time display and timezone.
 - Represent a directory with a (random) thumbnail.
 - Sorting of images and directories can be configured.
 - Tableless based layout for the gallery.
 - Images (and the cache) can be stored on different locations/(sub)domains on the
   server as long it's reachable by the browser.
 - Support for gif, jpg, jpeg, jpe, jif, jfif, jfi and png images.
 - Page navigation and breadcrumbs.
 - File information and output caching to speed everything up.
 - Reads Exif date from Jpeg images when possible (and use it for sorting).

Requirements
============

 - PHP 5.2
 - GD
 - Exif (optional)

Exif is most likely enabled in most PHP installations. See the PHP manual for
more information: http://php.net/manual/en/exif.installation.php

Installation
============

 - Drop the files from this package into a web accessible directory.
 - Make two directories named "images" and "cache" in the ImgBrowz0r directory.
 - Make "cache" directory writable (linux command: `chmod -R 0777 cache`).
 - Drop some images and directories in "images" directory.

That's a really basic setup. It's recommended to play with the configuration
and try various things (url rewriting, CMS integration, etc.).

Contact/Support
===============

Project page: http://61924.nl/projects/imgbrowz0r.html
Issues: http://github.com/FSX/imgbrowz0r/issues
Demo: http://images.61924.nl/

Development
===========

You can take a look at the development progress here:
http://github.com/FSX/imgbrowz0r

And bugs can be reported here:
http://github.com/FSX/imgbrowz0r/issues

Extra information
=================

Information about certain things.

  - Character blacklist

    Certain characters are not allowed in directory names.

    < > " ' & ; .. %

  - read_thumb_limit

    When the "random thumbnail" function is enabled it reads the cache of the current
    category/directory to choose a random thumbnail. If there are a lot thumbnails in
    the cache it will take more time to read all the files in the cache.

    You can set this to 10 to speedup the process, but it will only choose a random
    thumbnail of out the 10 first files it reads from the thumbnail cache.

  - Description for directories

    With imgbrowz0r::description() you can search for a description in the currently
    viewed directory. The function return the content of the description file when it
    found it and "null" is returned when nothing is found.

    You can add a description to a directory by adding a .desc file in the directory
    and put your description in that file.

    Html tags are stripped from the description except the following tags:
    <p>, <strong>, <em>, <a>, <br />, <h1>, <h2> and <h3>

     - Use valid HTML in your description to avoid unexpected behaviour.
     - Look for more info at:
       http://php.net/manual/en/function.strip-tags.php

  - Caching

    A small caching class, `ImgBrowz0rCache`, is provided with ImgBrowz0r. An instance
    of this class is passed to the ImgBrowz0r constructor to let ImgBrowz0r use the
    function of the caching class (see example.php for an example).

    This class can easily be replaced with an other cache as long as the other class
    has a public write and read method.

    A skeleton of of a caching class for ImgBrowz0r:

    class AlternativeImgBrowz0rCache
    {
        public function read($name)
        {
        }

        public function write($name, $data)
        {
        }
    }

    $name is the filename without the extension. The class must decide which extension
    to use. $data is an array. The read method must return the same array which was cached.


CHANGELOG
ImgBrowz0r changelog
http://61924.nl/projects/imgbrowz0r.html


0.3.7 - 22 July 2010:

 - Added "sort by EXIF DateTimeOriginal". Patch provided by bc.
 - Replaced scandir() with DirectoryIterator.
 - The timezone configuration option now accepts a valid TZ database timezone
   or `false`. When `false` is used the timezone setting will be skipped.
 - The enable_dst configuration option has been dropped.
 - Do not check if GD is enabled anymore.
 - Renamed the `imgbrowz0r` class to `ImgBrowz0r`.
 - Added small, optional, replaceable caching class.
 - A better pagination function.
 - Use P tags instead of DIV tags to wrap the breadcrumbs, pagination and statistics.
 - All files in the cache are now writeable by everyone. This ensures you can
   remove the cache files generated by the PHP/Apache user with your FTP client.
 - Added crop mode. This mode crops a random part of the image. (experimental)

TZ database timezones: http://en.wikipedia.org/wiki/List_of_tz_database_time_zones


0.3.6 - 16 February 2010:

 - Fixed directory traversal vulnerability (again). Reported by Secunia Research.


0.3.5 - 12 February 2010:

 - Renamed index.php to example.php.
 - Fixed directory traversal vulnerability. Fixed by bc.
 - *_sort_order options need different values. See example.php for information.
 - Some small optimizations (e.g. scandir instead of readdir).
 - Removed the filesize_limit option. This slowsdown everything and a user should
   be able to check his/her photos or other images before he puts them online.


0.3.4 - 10 September 2009:

 - Replaced str_replace() with preg_quote() for regex escaping.
 - Added separate sorting options for images and directories.
 - Made private methods/variables protected to make extending easier.
 - Added filesize limit option.
 - Added (preview) thumbnail generation for directories.
 - Changed markup for thumbnails. It'll be easier to style it with CSS.
 - Added memory peak usage to the example page (index.php). Now you can
   see how much memory GD (not ImgBrowz0r) uses.


0.3.3 - 29 June 2009:

 - Fixed a bug in the url generation for directories.
 - Directory names can now contain UTF-8 characters (tested with: 東京).
   Umlaut and stress and accent marks do work now.
 - Added character blacklist for directory names: < > " ' & ;


0.3.2 - 14 June 2009:

 - Breadcrumbs do not obey ['main_url']. Fixed by burnthewitches.


0.3.1 - 7 June 2009:

 - Added imgbrowz0r::description() to display a description. See README for more info.
 - Statistics are not shown in an empty directory.
 - Cleaned up code.


0.3 - 30 April 2009:

 - The functions that generate the navigation are removed from imgbrowz0r::display()
   and are now public functions.
 - Fixed a small bug in the url check part.
 - Added the random thumbnails again.
 - "Random thumbnails" can also be disabled now. Instead the same thumbnail will
   be shown each time or nothing at all.
 - Added a "read limit" that limits the amount of thumbnails that are used for the
   random thumbnail fucntion (this can speed things up).
 - Thumbnails are now refreshed if the orginal image has changed.
 - The supported file types are now stored in a variable.
 - Fixed a bug in imgbrowz0r::get_ext().
 - Fixed a bug in directory check in imgbrowz0r::make_thumb().
 - Directory statistics now in a  separate function, imgbrowz0r::statistics().
 - Removed check if GD extension is loaded. gd_info() function check is enough.
 - Removed imgbrowz0r::output_style().
 - Cleaned up page navigation generator.
 - Replaced some hard-coded values for the thumbnail height and width with variables
   in imgbrowz0r::make_thumb().


0.3-beta - 22 April 2009:

 - Most of it is rewritten.
 - Categories are called directories now.
 - Images can be dropped into the root gallery directory directly.
 - All thumbnails are now stored in a centralized cache.
 - The table based layout is replaced with a div based layout.
 - The sorting order of images and directories can now be set in the configuration.
 - The images and cache can now be placed on different locations/domains on a server.
 - Dropped base64 encoding for the category location in the url.
 - Improved quality of GIF thumbnails.
 - Improved page navigation.
 - Improved for environments that use url rewriting.
 - Added support for unlimited nesting of categories (directories).
 - Added support for daylight saving time (DST).
 - Added support for uppercase file extensions.
 - Added support for .jpeg, .jpe, .jif, .jfif and .jfi file extensions.
 - Added breadcrumbs navigation.
 - Added imgbrowz0r::output_style() to output a style for the div based layout.


0.2.2 - 26 September 2008:

 - Better file check.
 - Category names with non-latin characters now supported (see manual for more
   information). Patch provided by Serge Timakov.
 - Fixed bug in make_thumb(). Thumbnails of square images (images with dimensions
   like 600*600px) weren't resized.
 - Removed 'rename file if extension does not match image type' feature.
 - Added 'IMGBROWZ0R_VERSION' constant which contains the software version.


0.1.1 - 4 June 2008:

 - Use of gmdate() instead of date() in $this->format_time().
 - mkdir() now makes thumbs folder with chmod 0777 instead of 0700 in $this->make_thumb().
 - Fixed sorting bug in $this->view_categories(). Bug appeared when only one empty
   folder in the gallery folder.
 - $this->g_category is now false instead of empty when $_GET['c'] is ot set.
 - Removed unset()'s right before the return from $this->view_categories() and
   $this->view_images().
 - Made self::safe_url(), self::base64_url_encode(), self::base64_url_decode() and
   self::get_ext() static.


index.php
<?php

error_reporting(E_ALL);

// Some stuff for generating stats
function file_size($size)
{
	$units = array('b', 'Kb', 'Mb', 'Gb', 'Tb', 'Pb', 'Eb');

	for ($i = 0; $size > 1024; $i++)
		$size /= 1024;

	return round($size, 2).' '.$units[$i];
}

function get_microtime($microtime=false)
{
	if ($microtime === false)
		$microtime = microtime();

	list($usec, $sec) = explode(' ', $microtime);
	return ((float)$usec + (float)$sec);
}

$start_timer = microtime();

// Include class
require 'imgbrowz0r.php';

// These are all settings (set to default). The settings are not validated since you have to configure everything.
// There is a chance that ImgBrowz0r stops working if you enter the wrong values.
$config = array(
	// Directory settings. These are required. Without trailing slash. (required)
	'images_dir'               => '/var/www/imgbrowz0r/images',
	'cache_dir'                => '/var/www/imgbrowz0r/cache',

	// Url settings. These are required. Without trailing slash. (required)
	// %PATH% is replaced with the directory location and page number
	'main_url'                 => 'http://example.com/imgbrowz0r/index.php?q=%PATH%',
	'images_url'               => 'http://example.com/imgbrowz0r/images',
	'cache_url'                => 'http://example.com/imgbrowz0r/cache',

	// Sorting settings (optional)
	'dir_sort_by'              => 3, // 1 = filename, 2 = extension (dir), 3 = inode change time of file
	'dir_sort_order'           => false, // true = ascending, false = descending

	'img_sort_by'              => 3, // 1 = filename, 2 = extension (png, gif, etc.), 3 = inode change time of file
	'img_sort_order'           => false, // true = ascending, false = descending

	// Thumbnail settings (optional)
	'thumbs_per_page'          => 12, // Amount of thumbnails per page
	'max_thumb_row'            => 4, // Amount of thumbnails on a row
	'max_thumb_width'          => 200, // Maximum width of thumbnail
	'max_thumb_height'         => 200, // Maximum height of thumbnail

	// Time settings (optional)
	'time_format'              => 'F jS, Y', // Date formatting. Look at the PHP date() for help: http://nl3.php.net/manual/en/function.date.php
	'time_zone'                => 0, // Timezone. Example: 1
	'enable_dst'               => false, // Daylight saving time (DST). Set this to true to enable it.

	// Misc settings (optional)
	'ignore_port'              => false, // Ignore port in url. Set this to true to ignore the port.
	'dir_thumbs'               => false, // Show a thumbnail in a category box. Default is false.
	'random_thumbs'            => false, // Use random thumbnails for categories. Default is false.
	'read_thumb_limit'         => 0, // See README for information about this setting.
	'filesize_limit'           => 4069 // Filesize limit in kilobytes (4MB). Exclude images that are bigger than the filesize limit.
	);

// Start the class
$gallery = new imgbrowz0r($config);

// XHTML stuff
?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />

	<title>ImgBrowz0r 0.3.4</title>

	<style type="text/css">
		body { padding: 0;margin:0;text-align: center;font-size:90%;line-height: 1.5;background-color: #fff;font-family: Tahoma, Verdana, sans-serif }

		a:link, a:visited { color: #F52542;text-decoration: underline }
		a:hover, a:active { color: #333 }
		strong { font-weight: bold }

		/* This is some example CSS. You can change this to your own liking. */
		.img-description, .imgbrowz0r-navigation, #imgbrowz0r { margin: 0 auto;width: 970px;text-align: left }
		.imgbrowz0r-navigation { padding: 1.5em 0 0.5em }
		.imgbrowz0r-navigation .img-statistics { margin-top: 1.5em }

		.img-description { margin: 1.5em auto 0 }

		#imgbrowz0r .img-row { padding: 0 1em }

		#imgbrowz0r .img-directory span.img-dir-name,
		#imgbrowz0r .img-directory span.img-thumb-date { display: block }
		#imgbrowz0r .img-directory span.img-dir-name { font-weight: bold;font-size: 1.2em }

		#imgbrowz0r .img-column-1 { clear: left }

		#imgbrowz0r .img-thumbnail,
		#imgbrowz0r .img-directory { float: left;padding: 1.5em 0;width: 25%;text-align: center }

		#imgbrowz0r .img-thumbnail a,
		#imgbrowz0r .img-directory a {
			display: block;margin: 0 auto;
			width: 200px;height: 160px;line-height: 150px;
			background-repeat: no-repeat;background-position: 50% 50%; }

		#imgbrowz0r .img-directory a:link, #imgbrowz0r .img-directory a:visited, #imgbrowz0r .img-thumbnail a:link,
		#imgbrowz0r .img-thumbnail a:visited { background-color: #000;border: 1px solid #000;font-size: 1.5em;color: #ccc;text-decoration: none }
		#imgbrowz0r .img-directory a:active, #imgbrowz0r .img-directory a:hover, #imgbrowz0r .img-thumbnail a:active,
		#imgbrowz0r .img-thumbnail a:hover { border-color: #F52542;color: #F52542 }

		/* http://sonspring.com/journal/clearing-floats */
		html body div.clear,
		html body span.clear { background: none;border: 0;clear: both;display: block;float: none;font-size: 0;list-style: none;
				       margin: 0;padding: 0;overflow: hidden;visibility: hidden;width: 0;height: 0 }
	</style>
</head>
<body>

<?php

// Prepare everything. This function must be called before
// you call other functions. (required)
$gallery->init();

// Generate navigation and statistics. (optional, but remommended)
// The output of the functions are now assigned to variabled, but
// you can also call the functions directly.
$gallery_breadcrumbs = $gallery->breadcrumbs();
$gallery_pagination = $gallery->pagination();
$gallery_statistics = $gallery->statistics();

// Display description of the current directory. (optional)
echo $gallery->description();

// Display navigation
echo '<div class="imgbrowz0r-navigation">', "\n\t", $gallery_breadcrumbs, "\n\t", $gallery_pagination, "\n\t", $gallery_statistics, "\n", '</div>', "\n\n";

// Display images and directories. (required)
echo $gallery->browse();

// Display navigation
echo '<div class="imgbrowz0r-navigation">', "\n\t", $gallery_pagination, "\n\t", $gallery_breadcrumbs, "\n", '</div>', "\n\n";

// Showing some stats (optional)
echo '<p>Processing time: ', round(get_microtime(microtime()) - get_microtime($start_timer), 5),
	' &amp;&amp; Memory usage: ', file_size(memory_get_usage()),
	' &amp;&amp; Memory peak: ', file_size(memory_get_peak_usage()), '</p>';

?>

</body>
</html>


example.php
<?php

error_reporting(E_ALL);

// Some stuff for generating stats
function file_size($size)
{
	$units = array('b', 'Kb', 'Mb', 'Gb', 'Tb', 'Pb', 'Eb');

	for ($i = 0; $size > 1024; $i++)
		$size /= 1024;

	return round($size, 2).' '.$units[$i];
}

function get_microtime($microtime=false)
{
	if ($microtime === false)
		$microtime = microtime();

	list($usec, $sec) = explode(' ', $microtime);
	return ((float)$usec + (float)$sec);
}

$start_timer = microtime();

// Include class
require 'imgbrowz0r.php';

// These are all settings (set to default). The settings are not validated since you have to configure everything.
// There is a chance that ImgBrowz0r stops working if you enter the wrong values.
$config = array(
	// Directory settings. These are required. Without trailing slash. (required)
	'images_dir'               => dirname(__FILE__).'/images',
	'cache_dir'                => dirname(__FILE__).'/cache',

	// Url settings. These are required. Without trailing slash. (required)
	// %PATH% is replaced with the directory location and page number
	'main_url'                 => 'http://example.com/imgbrowz0r/example.php?q=%PATH%',
	'images_url'               => 'http://example.com/imgbrowz0r/images',
	'cache_url'                => 'http://example.com/imgbrowz0r/cache',

	// Sorting settings (optional)
	'dir_sort_by'              => 3, // 1 = filename, 2 = extension (dir), 3 = inode change time of file
	'img_sort_by'              => 3, // 1 = filename, 2 = extension (png, gif, etc.), 3 = inode change time of file or EXIF image data (Date Taken)

	// The sort order settings can have the following values:
	// SORT_ASC, SORT_DESC, SORT_REGULAR, SORT_NUMERIC, SORT_STRING
	// SORT_ASC = ascending, SORT_DESC = descending
	'dir_sort_order'           => SORT_DESC,
	'img_sort_order'           => SORT_DESC,

	// Thumbnail settings (optional)
	'thumbs_per_page'          => 12, // Amount of thumbnails per page
	'max_thumb_row'            => 4, // Amount of thumbnails on a row
	'max_thumb_width'          => 200, // Maximum width of thumbnail
	'max_thumb_height'         => 200, // Maximum height of thumbnail

	// Crop mode cuts out a random part of an image and uses it for the thumbnail
	'crop_mode'                => false,

	// Resize the image before cropping by 2.5. If the image mustn't be resized
	// the value can be set to 1.
	'crop_resize_factor'       => 2.5,

	// Date formatting. Look at the PHP date() for help: http://php.net/manual/en/function.date.php
	'time_format'              => 'F jS, Y',

	// Pick a valid timezone from http://en.wikipedia.org/wiki/List_of_tz_database_time_zones
	// Use `false` to disable the timezone option
	'time_zone'                => 'Europe/Amsterdam',

	// Daylight saving time (DST). Set this to true to enable it.
	'enable_dst'               => false,

	// Misc settings (optional)
	'ignore_port'              => false, // Ignore port in url. Set this to true to ignore the port.
	'dir_thumbs'               => false, // Show a thumbnail in a category box. Default is false.
	'random_thumbs'            => false, // Use random thumbnails for categories. Default is false.
	                                     // NOTE: random_thumbs will not work when the cache is enabled.
	'read_thumb_limit'         => 0 // See README for information about this setting.
	);

// Setup cache
$gallery_cache = new ImgBrowz0rCache(
	$config['cache_dir'], // The location of the cache directory. In this case the smae as imgBrowz0r's one.
	3600 // The amount of seconds the cache is valid.
);

// Start the class
$gallery = new ImgBrowz0r($config, $gallery_cache);

/* The cache is optional. If you don't want to use the cache you can remove
the "Setup cache" part and replace this:

    $gallery = new ImgBrowz0r($config, $gallery_cache);

With:

    $gallery = new ImgBrowz0r($config);

*/

// XHTML stuff
?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />

	<title>ImgBrowz0r <?php echo IMGBROWZ0R_VERSION ?></title>

	<style type="text/css">
		body {
		  padding: 0;
		  margin:0;
		  font-family: "Lucida Grande", "Lucida Sans Unicode", "Lucida Sans", Verdana, Arial, sans-serif;
		  text-align: center;
		  font-size: 87%;
		  line-height: 1.7em;
		  color: #222;
		  background-color: #fff }

		a:link, a:visited {
		  text-decoration: none;
		  font-weight: bold;
		  color: #EE3D40 }
		a:hover, a:active { color: black }
		strong { font-weight: bold }

		.pagination a { font-weight: normal }

		/* This is some example CSS. You can change this to your own liking. */

		.img-description, .imgbrowz0r-navigation, #imgbrowz0r {
		  margin: 0 auto;
		  width: 970px;
		  text-align: left }
		.imgbrowz0r-navigation { padding: 1.5em 0 0.5em }
		.imgbrowz0r-navigation .img-statistics { margin-top: 1.5em }

		.img-description { margin: 1.5em auto 0 }

		#imgbrowz0r .img-row { padding: 0 1em }

		#imgbrowz0r .img-directory span.img-dir-name,
		#imgbrowz0r .img-directory span.img-thumb-date { display: block }
		#imgbrowz0r .img-directory span.img-dir-name {
		  font-weight: bold;
		  font-size: 1.2em }

		#imgbrowz0r .img-column-1 { clear: left }

		#imgbrowz0r .img-thumbnail,
		#imgbrowz0r .img-directory {
		  float: left;
		  padding: 1.5em 0;
		  width: 25%;
		  text-align: center }

		#imgbrowz0r .img-thumbnail a,
		#imgbrowz0r .img-directory a {
		  display: block;
		  margin: 0 auto;
		  width: 200px;
		  height: 160px;
		  line-height: 150px;
		  background-repeat: no-repeat;
		  background-position: 50% 50% }

		#imgbrowz0r .img-directory a:link, #imgbrowz0r .img-directory a:visited,
		#imgbrowz0r .img-thumbnail a:link,
		#imgbrowz0r .img-thumbnail a:visited {
		  background-color: #000;
		  border: 3px solid #000;
		  overflow: hidden; /* If the title of the directory is too long the overflowing
		                       part will be hidden. Not a perfect solution. */
		  font-size: 1.5em;
		  color: #ccc;
		  text-decoration: none }
		#imgbrowz0r .img-directory a:active, #imgbrowz0r .img-directory a:hover,
		#imgbrowz0r .img-thumbnail a:active,
		#imgbrowz0r .img-thumbnail a:hover {
		  border-color: #F52542;
		  color: #F52542 }

		#imgbrowz0r .img-directory a, #imgbrowz0r .img-thumbnail a {
		  text-indent: -9999em }
		#imgbrowz0r .img-no-thumbnail a {
		  text-indent: 0 }

		/* http://sonspring.com/journal/clearing-floats */
		html body div.clear,
		html body span.clear {
		  background: none;
		  border: 0;
		  clear: both;
		  display: block;
		  float: none;
		  font-size: 0;
		  list-style: none;
		  margin: 0;
		  padding: 0;
		  overflow: hidden;
		  visibility: hidden;
		  width: 0;
		  height: 0 }
	</style>
</head>
<body>

<?php

// Prepare everything. This function must be called before
// you call other functions. (required)
$gallery->init();

// Generate navigation and statistics. (optional, but remommended)
// The output of the functions are now assigned to variabled, but
// you can also call the functions directly.
$gallery_breadcrumbs = $gallery->breadcrumbs();
$gallery_pagination = $gallery->pagination();
$gallery_statistics = $gallery->statistics();

// Display description of the current directory. (optional)
echo $gallery->description();

// Display navigation
echo '<div class="imgbrowz0r-navigation">', "\n\t",
     $gallery_breadcrumbs, "\n\t",
	 $gallery_pagination, "\n\t",
	 $gallery_statistics, "\n", '</div>', "\n\n";

// Display images and directories. (required)
echo $gallery->browse();

// Display navigation
echo '<div class="imgbrowz0r-navigation">', "\n\t",
     $gallery_pagination, "\n\t",
	 $gallery_breadcrumbs, "\n", '</div>', "\n\n";

// Showing some stats (optional)
echo '<p>Processing time: ', round(get_microtime(microtime()) - get_microtime($start_timer), 5),
	' &amp;&amp; Memory usage: ', file_size(memory_get_usage()),
	' &amp;&amp; Memory peak: ', file_size(memory_get_peak_usage()), '</p>';

?>

</body>
</html>
